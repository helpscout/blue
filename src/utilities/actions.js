export const createActionTypes = (actionTypes = [], namespace = '@@HSDS') => {
  return actionTypes.reduce((types, type) => {
    return { ...types, [type]: `${namespace}/${type}` }
  }, {})
}
