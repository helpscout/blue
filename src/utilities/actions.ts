export const createActionTypes = (
  actionTypes: Array<string> = [],
  namespace: string = '@@HSDS'
): any => {
  return actionTypes.reduce((types, type) => {
    return { ...types, [type]: `${namespace}/${type}` }
  }, {})
}
