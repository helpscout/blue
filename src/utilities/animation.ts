import { isArray, isString } from './is'

export type Sequence = string
export type SequenceNames = Array<Sequence>
export type AnimationSequence = SequenceNames | Sequence

export const getSequenceNames = (
  sequence: AnimationSequence
): SequenceNames => {
  let names = <SequenceNames>[]

  if (isArray(sequence)) {
    names = <SequenceNames>sequence
  }

  if (isString(sequence)) {
    names = (<Sequence>sequence)
      .trim()
      .split(' ')
      .filter(n => n !== '')
  }

  return names
}
