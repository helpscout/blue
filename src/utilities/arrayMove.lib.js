/**
 * Do not convert this file to TypeScript!
 * This exists to ensure that the imports work correctly.
 */
import arrayMove from 'array-move'

export default arrayMove
