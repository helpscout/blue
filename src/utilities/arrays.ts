/**
 * Returns the first item of an array
 * @param  {Array} array The array.
 * @returns {any}
 */
export const first = (array: Array<any> = []): any => array[0]

/**
 * Returns the last item of an array
 * @param   {Array} array The array.
 * @returns {any}
 */
export const last = (array: Array<any> = []): any => array[array.length - 1]

/**
 * Returns a random item from an array
 *
 * @param   {Array} array The array.
 * @returns {any}
 */
export const random = (array: Array<any> = []): any => {
  return array[Math.floor(Math.random() * array.length)]
}

/**
 * Determines if an array contains an item. Small polyfill for Array.includes.
 *
 * @param   {Array} array The array.
 * @param   {any} item The item to match.
 * @returns {boolean} The result.
 */
export const includes = (array: Array<any> = [], item: any): boolean => {
  return array.indexOf(item) >= 0
}
