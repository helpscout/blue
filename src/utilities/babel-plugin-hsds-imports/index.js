'use strict'
const { readdirSync, statSync } = require('fs')
const { join, resolve } = require('path')

const dirs = p => readdirSync(p).filter(f => statSync(join(p, f)).isDirectory())
let COMPONENT_NAMES = []

try {
  COMPONENT_NAMES = dirs(resolve(__dirname, '../../components'))
} catch (e) {
  throw new Error('Components dir not found')
}

module.exports = function(babel) {
  const { types: t } = babel

  return {
    name: 'hsds-component-import',
    visitor: {
      ImportDeclaration(path, state) {
        const packages = state.opts.packages
          ? [].concat(state.opts.packages)
          : ['@helpscout/hsds-react']
        let packageName = path.node.source.value

        if (isWrongPackage(packageName, packages)) return null
        if (!areAllValidComponents(path.node.specifiers)) return null

        if (packageName.endsWith('/')) {
          packageName = packageName.slice(0, packageName.length - 1)
        }

        if (!packageName.includes('components')) {
          packageName += '/components'
        }

        const program = path.findParent(t.isProgram)

        path.node.specifiers.forEach(specifier => {
          const identifier = specifier.local
          const identifierName = identifier.name
          const defaultImportSpecifier = t.importDefaultSpecifier(identifier)
          let pkgPath = `${packageName}`

          if (!packageName.includes(identifierName)) {
            pkgPath += `/${identifierName}`
          }

          const newImportDeclaration = t.importDeclaration(
            [defaultImportSpecifier],
            t.stringLiteral(pkgPath)
          )
          program.node.body.unshift(newImportDeclaration)
        })

        path.remove()
      },
    },
  }
}

function areAllValidComponents(specifiers) {
  if (!specifiers.length) return false
  if (!COMPONENT_NAMES.length) return false
  // If Button is not present something odd happened, bail
  if (!COMPONENT_NAMES.includes('Button')) return false

  for (let i = 0; i < specifiers.length; i++) {
    if (!COMPONENT_NAMES.includes(specifiers[i].local.name)) {
      return false
    }
  }

  return true
}

function isWrongPackage(packageName, packages) {
  if (!packageName) return true

  let found = false

  for (let i = 0; i < packages.length; i++) {
    if (packageName.includes(packages[i])) {
      found = true
    }
  }

  return !found
}
