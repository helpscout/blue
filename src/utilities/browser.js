/* istanbul ignore next */
// Can't write tests for this in JSDOM.
// Can't create fixture for JSDOM's built-in Navigator instance.
export const isBrowser = browser => {
  if (!navigator) return false
  return navigator.userAgent.toLowerCase().indexOf(browser) > -1
}

export const isEdge = () => isBrowser('edge')
export const isChrome = () => isBrowser('chrome')
export const isFirefox = () => isBrowser('firefox')
export const isSafari = () => isBrowser('safari')
