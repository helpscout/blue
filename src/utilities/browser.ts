type Browser = string

/* istanbul ignore next */
// Can't write tests for this in JSDOM.
// Can't create fixture for JSDOM's built-in Navigator instance.
export const isBrowser = (browser: Browser) => {
  if (!navigator) return false
  return navigator.userAgent.toLowerCase().indexOf(browser) > -1
}

export const isEdge = (): boolean => isBrowser('edge')
export const isChrome = (): boolean => isBrowser('chrome')
export const isFirefox = (): boolean => isBrowser('firefox')
export const isSafari = (): boolean => isBrowser('safari')
