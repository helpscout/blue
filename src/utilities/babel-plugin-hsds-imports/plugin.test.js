import pluginTester from 'babel-plugin-tester'
import hsdsImports from './'

pluginTester({
  plugin: hsdsImports,
  snapshot: true,
  tests: [
    {
      title: 'Should not change imports that are not hsds or in the options',
      code: `
      import React from 'react'
      import _ from 'lodash'
      `,
    },
    {
      title: 'Should not change default imports',
      code: `
      import Page from '@helpscout/hsds-react/components/Page'
      `,
    },
    {
      title:
        'Should change destructured hsds imports to default style imports (1 item)',
      code: `
      import { Tooltip } from '@helpscout/hsds-react/components/'
      `,
    },
    {
      title:
        'Should change destructured hsds imports to default style imports (2+ items)',
      code: `
      import { Tooltip, Select } from '@helpscout/hsds-react/components/'
      `,
    },
    {
      title:
        'Should make sure that "components" is in the transformed path in destructured imports',
      code: `
      import { Tooltip, Select } from '@helpscout/hsds-react'
      `,
    },
    {
      title:
        'Should make sure that "components" is in the transformed path in default imports',
      code: `
      import Button from '@helpscout/hsds-react'
      `,
    },
    {
      title: 'Should not change on inexistent/invalid components',
      code: `
      import { Hello } from '@helpscout/hsds-react'
      `,
    },
    {
      title:
        'Should not change on inexistent/invalid components on a multiple import',
      code: `
      import { Button, Hello } from '@helpscout/hsds-react'
      `,
    },
    {
      title: 'Should not change other imports',
      code: `
      import { OverlayUI } from '@helpscout/hsds-react/components/Overlay/Overlay.css'
      import Keys from '@helpscout/hsds-react/constants/Keys'
      `,
    },
    {
      title: 'Should not change HSDS utilities',
      code: `
      import { getColor } from '@helpscout/hsds-react/styles/utilities/color'
      `,
    },
    {
      title: 'Should not change HSDS adapters',
      code: `
      import '@helpscout/hsds-react/adapters/app'
      `,
    },
    {
      title: 'Should work with mixed stuff',
      code: `
      import Button from '@helpscout/hsds-react'
      import Page from '@helpscout/hsds-react/components/Page'
      import { Tooltip, Select } from '@helpscout/hsds-react/components/'
      import React from 'react'
      import { getColor } from '@helpscout/hsds-react/styles/utilities/color'
      `,
    },
    {
      title: 'Should work with options',
      pluginOptions: {
        packages: ['hsds-react-next'],
      },
      code: `
      import Button from 'hsds-react-next'
      import Page from 'hsds-react-next/components/Page'
      `,
    },
    {
      title: 'Should work with options (single)',
      pluginOptions: {
        packages: 'hsds-react-next',
      },
      code: `
      import Button from 'hsds-react-next'
      import Page from 'hsds-react-next/components/Page'
      `,
    },
    {
      title: 'Should work with options (2+)',
      pluginOptions: {
        packages: ['hsds-react-next', '@helpscout/hsds-react'],
      },
      code: `
      import Button from 'hsds-react-next'
      import Page from 'hsds-react-next/components/Page'
      import { Tooltip, Select } from '@helpscout/hsds-react/components/'
      `,
    },
  ],
})
