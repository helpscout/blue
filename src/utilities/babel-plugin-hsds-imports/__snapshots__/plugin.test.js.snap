// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hsds-component-import Should change destructured hsds imports to default style imports (1 item): Should change destructured hsds imports to default style imports (1 item) 1`] = `

import { Tooltip } from '@helpscout/hsds-react/components/'

      ↓ ↓ ↓ ↓ ↓ ↓

import Tooltip from '@helpscout/hsds-react/components/Tooltip'


`;

exports[`hsds-component-import Should change destructured hsds imports to default style imports (2+ items): Should change destructured hsds imports to default style imports (2+ items) 1`] = `

import { Tooltip, Select } from '@helpscout/hsds-react/components/'

      ↓ ↓ ↓ ↓ ↓ ↓

import Select from '@helpscout/hsds-react/components/Select'
import Tooltip from '@helpscout/hsds-react/components/Tooltip'


`;

exports[`hsds-component-import Should make sure that "components" is in the transformed path in default imports: Should make sure that "components" is in the transformed path in default imports 1`] = `

import Button from '@helpscout/hsds-react'

      ↓ ↓ ↓ ↓ ↓ ↓

import Button from '@helpscout/hsds-react/components/Button'


`;

exports[`hsds-component-import Should make sure that "components" is in the transformed path in destructured imports: Should make sure that "components" is in the transformed path in destructured imports 1`] = `

import { Tooltip, Select } from '@helpscout/hsds-react'

      ↓ ↓ ↓ ↓ ↓ ↓

import Select from '@helpscout/hsds-react/components/Select'
import Tooltip from '@helpscout/hsds-react/components/Tooltip'


`;

exports[`hsds-component-import Should not change HSDS adapters: Should not change HSDS adapters 1`] = `

import '@helpscout/hsds-react/adapters/app'

      ↓ ↓ ↓ ↓ ↓ ↓

import '@helpscout/hsds-react/adapters/app'


`;

exports[`hsds-component-import Should not change HSDS utilities: Should not change HSDS utilities 1`] = `

import { getColor } from '@helpscout/hsds-react/styles/utilities/color'

      ↓ ↓ ↓ ↓ ↓ ↓

import { getColor } from '@helpscout/hsds-react/styles/utilities/color'


`;

exports[`hsds-component-import Should not change default imports: Should not change default imports 1`] = `

import Page from '@helpscout/hsds-react/components/Page'

      ↓ ↓ ↓ ↓ ↓ ↓

import Page from '@helpscout/hsds-react/components/Page'


`;

exports[`hsds-component-import Should not change imports that are not hsds or in the options: Should not change imports that are not hsds or in the options 1`] = `

import React from 'react'
import _ from 'lodash'

      ↓ ↓ ↓ ↓ ↓ ↓

import React from 'react'
import _ from 'lodash'


`;

exports[`hsds-component-import Should not change on inexistent/invalid components on a multiple import: Should not change on inexistent/invalid components on a multiple import 1`] = `

import { Button, Hello } from '@helpscout/hsds-react'

      ↓ ↓ ↓ ↓ ↓ ↓

import { Button, Hello } from '@helpscout/hsds-react'


`;

exports[`hsds-component-import Should not change on inexistent/invalid components: Should not change on inexistent/invalid components 1`] = `

import { Hello } from '@helpscout/hsds-react'

      ↓ ↓ ↓ ↓ ↓ ↓

import { Hello } from '@helpscout/hsds-react'


`;

exports[`hsds-component-import Should not change other imports: Should not change other imports 1`] = `

import { OverlayUI } from '@helpscout/hsds-react/components/Overlay/Overlay.css'
import Keys from '@helpscout/hsds-react/constants/Keys'

      ↓ ↓ ↓ ↓ ↓ ↓

import { OverlayUI } from '@helpscout/hsds-react/components/Overlay/Overlay.css'
import Keys from '@helpscout/hsds-react/constants/Keys'


`;

exports[`hsds-component-import Should work with mixed stuff: Should work with mixed stuff 1`] = `

import Button from '@helpscout/hsds-react'
import Page from '@helpscout/hsds-react/components/Page'
import { Tooltip, Select } from '@helpscout/hsds-react/components/'
import React from 'react'
import { getColor } from '@helpscout/hsds-react/styles/utilities/color'

      ↓ ↓ ↓ ↓ ↓ ↓

import Select from '@helpscout/hsds-react/components/Select'
import Tooltip from '@helpscout/hsds-react/components/Tooltip'
import Page from '@helpscout/hsds-react/components/Page'
import Button from '@helpscout/hsds-react/components/Button'
import React from 'react'
import { getColor } from '@helpscout/hsds-react/styles/utilities/color'


`;

exports[`hsds-component-import Should work with options (2+): Should work with options (2+) 1`] = `

import Button from 'hsds-react-next'
import Page from 'hsds-react-next/components/Page'
import { Tooltip, Select } from '@helpscout/hsds-react/components/'

      ↓ ↓ ↓ ↓ ↓ ↓

import Select from '@helpscout/hsds-react/components/Select'
import Tooltip from '@helpscout/hsds-react/components/Tooltip'
import Page from 'hsds-react-next/components/Page'
import Button from 'hsds-react-next/components/Button'


`;

exports[`hsds-component-import Should work with options (single): Should work with options (single) 1`] = `

import Button from 'hsds-react-next'
import Page from 'hsds-react-next/components/Page'

      ↓ ↓ ↓ ↓ ↓ ↓

import Page from 'hsds-react-next/components/Page'
import Button from 'hsds-react-next/components/Button'


`;

exports[`hsds-component-import Should work with options: Should work with options 1`] = `

import Button from 'hsds-react-next'
import Page from 'hsds-react-next/components/Page'

      ↓ ↓ ↓ ↓ ↓ ↓

import Page from 'hsds-react-next/components/Page'
import Button from 'hsds-react-next/components/Button'


`;
